File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/InviteGroupRequest.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.UUID;

/**
 * InviteGroupRequest
 * viewer -> simulator -> dataserver
 * reliable
 *
 * <p>Template: {@code InviteGroupRequest Low 349 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class InviteGroupRequest extends SLMessage {
    public AgentData AgentData_Field;
    public GroupData GroupData_Field;
    public ArrayList<InviteData> InviteData_Fields = new ArrayList<>();

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID - UUID of inviting agent
        public UUID SessionID; // LLUUID
    }

    /** Block GroupData, Single. */
    public static class GroupData {
        public UUID GroupID; // LLUUID
    }

    /** Block InviteData, Variable. */
    public static class InviteData {
        public UUID InviteeID; // LLUUID
        public UUID RoleID; // LLUUID
    }

    public InviteGroupRequest() {
        this.zeroCoded = false;
        this.AgentData_Field = new AgentData();
        this.GroupData_Field = new GroupData();
    }

    @Override
    public int CalcPayloadSize() {
        return (this.InviteData_Fields.size() * 32) + 53;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleInviteGroupRequest(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 349 (InviteGroupRequest).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x01);
        byteBuffer.put((byte) 0x5D);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packUUID(byteBuffer, this.GroupData_Field.GroupID);
        byteBuffer.put((byte) this.InviteData_Fields.size());
        for (InviteData inviteData : this.InviteData_Fields) {
            packUUID(byteBuffer, inviteData.InviteeID);
            packUUID(byteBuffer, inviteData.RoleID);
        }
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.GroupData_Field.GroupID = unpackUUID(byteBuffer);
        int i = byteBuffer.get() & 0xFF;
        for (int j = 0; j < i; j++) {
            InviteData inviteData = new InviteData();
            inviteData.InviteeID = unpackUUID(byteBuffer);
            inviteData.RoleID = unpackUUID(byteBuffer);
            this.InviteData_Fields.add(inviteData);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/InviteGroupRequest.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.UUID

/**
 * InviteGroupRequest
 * viewer -> simulator -> dataserver
 * reliable
 *
 * <p>Template: {@code InviteGroupRequest Low 349 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class InviteGroupRequest : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var GroupData_Field: GroupData = GroupData()
    @JvmField val InviteData_Fields = ArrayList<InviteData>()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block GroupData, Single. */
    open class GroupData {
        @JvmField var GroupID: UUID? = null
    }

    /** Block InviteData, Variable. */
    open class InviteData {
        @JvmField var InviteeID: UUID? = null
        @JvmField var RoleID: UUID? = null
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return (InviteData_Fields.size * 32) + 53
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleInviteGroupRequest(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 349 (InviteGroupRequest).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x01).toByte())
        byteBuffer.put((0x5D).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packUUID(byteBuffer, GroupData_Field.GroupID)
        byteBuffer.put((InviteData_Fields.size.toByte()))
        for (inviteData in InviteData_Fields) {
            packUUID(byteBuffer, inviteData.InviteeID)
            packUUID(byteBuffer, inviteData.RoleID)
        }
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        GroupData_Field.GroupID = unpackUUID(byteBuffer)
        val i = (byteBuffer.get().toInt() and 0xFF)
        repeat(i) {
            val inviteData = InviteData()
            inviteData.InviteeID = unpackUUID(byteBuffer)
            inviteData.RoleID = unpackUUID(byteBuffer)
            InviteData_Fields.add(inviteData)
        }
    }
}
